### PR TITLE
Implement pending bits for plic

### PR DESCRIPTION
**Pull request by jiegec**
Fix issue #1280

### CODE REVIEW

**Comment by aswaterman** [comment]
cc @avpatel

**Review by scottj97 — APPROVED** [review]
I'm not that familiar with PLIC; I assume these are supposed to be read-only? Since you didn't touch `plic_t::store`.

**Comment by aswaterman** [comment]
@avpatel can you please review?

File: riscv/plic.cc [hunk 1]
```diff
@@ -313,8 +316,16 @@ bool plic_t::load(reg_t addr, size_t len, uint8_t* bytes)
       return false;
   }
 
-  if (PRIORITY_BASE <= addr && addr < ENABLE_BASE) {
+  if (PRIORITY_BASE <= addr && addr < PENDING_BASE) {
     ret = priority_read(addr, &val);
+  } else if (PENDING_BASE <= addr && addr < ENABLE_BASE) {
+    uint32_t id_word = (addr - PENDING_BASE) >> 2;
```
Comment (avpatel) [inline riscv/plic.cc]: To be consistent with other parts of PLIC emulation, add `bool plic_t::pending_read(reg_t offset, uint32_t *val)` and use it here.

File: riscv/plic.cc [hunk 2]
```diff
@@ -313,8 +316,16 @@ bool plic_t::load(reg_t addr, size_t len, uint8_t* bytes)
...
+    if (id_word < num_ids_word) {
+      val = level[id_word];
```
Comment (avpatel) [inline riscv/plic.cc]: This is not correct. The pending register should be an ORing of pending state in each context

```
val = 0;
for (i = 0; i < contexts.size(); i++) {
    val |= contexts.pending[id_word];
}
```

File: riscv/plic.cc (hunk 1 above)
Comment (scottj97) [reply riscv/plic.cc]: I assume you meant `pending_read()`?

Comment (avpatel) [reply riscv/plic.cc]: Ahh, yes. It was a typo. I have fixed it.

Comment (jiegec) [reply riscv/plic.cc]: Done.

File: riscv/plic.cc (hunk 2 above)
Comment (jiegec) [reply riscv/plic.cc]: Done.

File: riscv/plic.cc [hunk 3]
```diff
@@ -156,6 +159,21 @@ bool plic_t::priority_write(reg_t offset, uint32_t val)
   return true;
 }
 
+bool plic_t::pending_read(reg_t offset, uint32_t *val)
+{
+  uint32_t id_word = (offset >> 2);
+
+  if (id_word > 0 && id_word < num_ids_word) {
```
Comment (avpatel) [inline riscv/plic.cc]: Only `id_word < num_ids_word` check is required here.

**Comment by avpatel** [comment]
@jiegec Only minor comment, otherwise this looks good.

File: riscv/plic.cc (hunk 3 above)
Comment (jiegec) [reply riscv/plic.cc]: Thanks!

**Comment by avpatel** [comment]
@aswaterman This looks good to me now. Please merge it.

File: riscv/plic.cc [hunk 4]
```diff
@@ -156,6 +159,21 @@ bool plic_t::priority_write(reg_t offset, uint32_t val)
   return true;
 }
 
+bool plic_t::pending_read(reg_t offset, uint32_t *val)
+{
+  uint32_t id_word = (offset >> 2);
+
+  if (id_word < num_ids_word) {
+    *val = 0;
+    for (size_t i = 0; i < contexts.size(); i++) {
```
Comment (scottj97) [inline riscv/plic.cc]: Please use a modern C++ loop: `for (auto context: contexts)`

**Review by scottj97 — CHANGES_REQUESTED** [review]

File: riscv/plic.cc (hunk 4 above)
Comment (jiegec) [reply riscv/plic.cc]: Done

**Review by scottj97 — APPROVED** [review]

**Review by aswaterman — APPROVED** [review]

**Pull request merged**